pulsar-function-go/pf/instance.go

```diff
@@ -0,0 +1,256 @@
+		subscriptionType = pulsar.Failover
+	}
+
+	funcDetails := gi.context.InstanceConf.FuncDetails
+	subscriptionName := funcDetails.Tenant + "/" + funcDetails.Namespace + "/" + funcDetails.Name
+
+	properties := getProperties(getDefaultSubscriptionName(
+		funcDetails.Tenant,
+		funcDetails.Namespace,
+		funcDetails.Name), gi.context.InstanceConf.InstanceID)
+
+	channel := make(chan pulsar.ConsumerMessage)
+
+	var (
+		consumer pulsar.Consumer
+		err      error
+	)
+
+	for topic, consumerConf := range funcDetails.Source.InputSpecs {
+		log.Debugf("Setting up consumer for topic: %s with subscription name: %s", topic, subscriptionName)
+		if consumerConf.ReceiverQueueSize != nil {
+			if consumerConf.IsRegexPattern {
+				consumer, err = gi.client.Subscribe(pulsar.ConsumerOptions{
+					TopicsPattern:     topic,
+					ReceiverQueueSize: int(consumerConf.ReceiverQueueSize.Value),
+					SubscriptionName:  subscriptionName,
+					Properties:        properties,
+					Type:              subscriptionType,
+					MessageChannel:    channel,
+				})
+			} else {
+				consumer, err = gi.client.Subscribe(pulsar.ConsumerOptions{
+					Topic:             topic,
+					SubscriptionName:  subscriptionName,
+					Properties:        properties,
+					Type:              subscriptionType,
+					ReceiverQueueSize: int(consumerConf.ReceiverQueueSize.Value),
+					MessageChannel:    channel,
+				})
+			}
+		} else {
+			if consumerConf.IsRegexPattern {
+				consumer, err = gi.client.Subscribe(pulsar.ConsumerOptions{
+					TopicsPattern:    topic,
+					SubscriptionName: subscriptionName,
+					Properties:       properties,
+					Type:             subscriptionType,
+					MessageChannel:   channel,
+				})
+			} else {
+				consumer, err = gi.client.Subscribe(pulsar.ConsumerOptions{
+					Topic:            topic,
+					SubscriptionName: subscriptionName,
+					Properties:       properties,
+					Type:             subscriptionType,
+					MessageChannel:   channel,
+				})
+
+			}
+		}
+
+		if err != nil {
+			log.Errorf("create consumer error:%s", err.Error())
+		}
+		gi.consumers[topic] = consumer
+	}
+	return channel
+}
+
+func (gi *GoInstance) handlerMsg(input pulsar.Message) (output []byte, err error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+	ctx = NewContext(ctx, gi.context)
+	msgInput := input.Payload()
+	output, err = gi.function.Process(ctx, msgInput)
+	if err != nil {
+		log.Errorf("process function err:%v", err)
+		return nil, err
+	}
+
+	return output, nil
+}
+
+func (gi *GoInstance) processResult(msgInput pulsar.Message, output []byte) error {
+	atMostOnce := gi.context.InstanceConf.FuncDetails.ProcessingGuarantees == pb.ProcessingGuarantees_ATMOST_ONCE
+	atLeastOnce := gi.context.InstanceConf.FuncDetails.ProcessingGuarantees == pb.ProcessingGuarantees_ATLEAST_ONCE
+	autoAck := gi.context.InstanceConf.FuncDetails.AutoAck
+
+	if atMostOnce && autoAck {
+		for topic, consumer := range gi.consumers {
```

Comment:
introduce two methods: `ackInputMessage` and `nackInputMessage`.

```
func ackInputMessage(inputMessage: Message) {
    gi.consumers[inputMessage.Topic()].Ack(inputMessage)
}
```

```
func nackInputMessage(inputMessage: Message) {
    gi.consumers[inputMessage.Topic()].Nack(inputMessage)
}
```

so the logic here can be simplified into one method call:

```
ackInputMessage(msgInput)
```

And you don't need to handle the error for Ack.